### ISSUE TITLE
feature request: new bazel query function to help query macros

### ISSUE THREAD

**Issue by dgoldstein0** [bazelbuild/bazel#11848]
### Description of the problem / feature request:

Bazel query today operates at a target level.  It helpfully also includes generator_name, generator_function, and generator_location attributes for targets generated from a macro; but there's no way to effectively filter rules generated by macros.

In particular, I was writing a query to find dead targets in our codebase, which was roughly (vastly simplified version) `kind(rule, //...) - deps(kind(.*_binary, //...) + tests(//...))`; the assumption here being that every binary and test target were used, and anything not transitive deps of those, is not.  Those assumptions aren't perfect, but they are good enough for now.

One of the problems I ran into, however, is that I didn't want to consider some targets generated by a macro dead if others weren't.  For example, we have a spark macro that generates a binary and some metadata targets, and we deploy all of these together; but the binary it generates doesn't depend on the metadata.  Whether this setup is ideal, I didn't dig in to give an opinion on that.  What I wanted was for my query to ignore the metadata targets, since one of the targets generated from the macro was considered used (the binary).

### Feature requests: what underlying problem are you trying to solve with this feature?

For this particular use case, I think a new bazel query operator would make sense: `same_package_same_generator_name` (name could probably use work).  I.e. if there's a build file like

```
# package //foo

whatever_macro( # generates ":bar", "bar_foo", "bar_baz"
  name = "bar"
  ...
)
```

then `bazel query "same_package_same_generator_name(//foo:bar)"`, `bazel query "same_package_same_generator_name(//foo:bar_foo)"`, and `bazel query "same_package_same_generator_name(//foo:bar_baz)"` would all evaluate to

```
//foo:bar
//foo:bar_baz
//foo:bar_foo
```

For targets without a `generator_name`, this function would instead behave as the identity function.

Then my query for dead targets could be

```
kind(rule, //...) - same_package_same_generator_function(deps(kind(.*_binary, //...) + tests(//...)))
```

and this way it would consider all targets of a macro "live" if any of the targets generated by the macro are used.

### What operating system are you running Bazel on?

Linux

### What's the output of `bazel info release`?

release 3.3.0-5dd95469d237592d70e55fcc0e029f24bc07516b

### What's the output of `git remote get-url origin ; git rev-parse master ; git rev-parse HEAD` ?

not relevant

###  Have you found anything relevant by searching the web?

I read through the bazel query how-to and reference, but couldn't find anything useful for this use case.

### Any other information, logs, or outputs that you want to share?

nope

**Comment by brandjon** [comment]
Rather than add new query operators, I think this use case can be covered by our broader plans to export a much more detailed and useful version of Bazel's loading-phase information. This would include the target graph as well as things like rule declarations and the .bzl load DAG. You could then write a separate script to consume this data and run your query.

(No design doc available yet.)

**Comment by dgoldstein0** [comment]
that sounds quite a bit more complicated to use than just writing a single `bazel query`.

One of the big problems I see with that interface is that we currently have a service that can execute queries remotely, quite a bit faster than I can execute the same queries locally.  If the macro information isn't exposed via bazel query, I can no longer use that service for acceleration - I'd have to run the loading phase locally.  which probably also means running the fetching phase locally, which for us is quite slow.

**Comment by brandjon** [comment]
Not necessarily. You could have the remote service send the proto output back to you quickly (potentially even streaming it), then run the analysis locally.

But there are some downsides. First, you're paying the cost of dumping all the detailed information needed to run the post-processing, whereas Bazel already has that information internally. Second, if it's not a built-in operator, you can't compose your filtering step with other query operators. (You could of course reimplement all of Bazel query's operators in the post-processing, but once again that precludes any eager filtering.)

So perhaps if this feature is useful to many users, it warrants its own operator.

**Comment by dgoldstein0** [comment]
My vote is certainly for new query operator.  Any changes to our remote service could be organizationally difficult, as my team doesn't own that and it'd be quite a change in the interface for that service; making it a query operator makes the work I need to use it with our remote service minutes instead of days or weeks.

**Comment by dgoldstein0** [comment]
I am still waiting for this feature.  If something like this is currently doable in the bazel query language I'm open to using that instead, but to my knowledge there's no way to write "expand this set of targets I've found to all with a matching generator_name" (or any other such attribute match).

**Comment by stagnation** [comment]
I think the "meta attribute" can be used in the `attr` filter:

```
$ bazel query 'attr(generator_function, py_binary, //...)'
py_binary rule //:Runner
py_binary rule //Parameters:Generate
py_binary rule //config:Runner
$ bazel query 'kind(py_binary, //...)'
py_binary rule //:Runner
py_binary rule //Parameters:Generate
py_binary rule //config:Runner
```

**Comment by dgoldstein0** [comment]
That's true but I don't think it helps.  I want to query "expand this set
to any targets in the same package where the generator_name matches".
attr() itself helps me filter an existing set of targets I've found, not
expand it.

**Comment by stagnation** [comment]
Aha, that makes sense, I was too hasty in replying to comprehend the ask fully :)
Your use case is more complex than what I've had. Maybe some of the set-operations and staged queries would do,
or just a post-processing, database-like join would be best. Or, as you say a query function if it is easy from the canonical representation, rather than have consumers recreate the graph and do their own queries.

**Comment by dgoldstein0** [comment]
I'm still interested in a fix for this
